### ISSUE TITLE
[Unity] Bug: Number overloads

### ISSUE THREAD

**Issue by raad-genies** [Tencent/puerts#1871]
### 前置阅读 | Pre-reading

- [X] [文档](https://puerts.github.io/unity/zhcn/readme)/[Doc](https://puerts.github.io/unity/en/readme)
- [X] [Issues](https://github.com/Tencent/puerts/issues?q=is%3Aissue+sort%3Aupdated-desc+is%3Aclosed)
- [X] [FAQ](https://github.com/Tencent/puerts/blob/master/doc/unity/zhcn/faq.md)

### Puer的版本 | Puer Version

Latest

### Unity的版本 | Unity Version

2022.3

### 发生在哪个平台 | Platform

Editor(mac)

### 错误信息 | Error Message

Not an error but more of an issue with how number overloads work. In the case of `Random.Range` Puerts currently has no way of distinguishing between the float and the int overloads which results in float always passing since number can always be a float. 

Also any tips around how to create the same overload in the TS side would be great. Branding can work but then its essentially a new type and can't really just say `var x : float = 9` for example. 

### 问题重现 | Bug reproduce

Bind Random.Range and try to call it with decimals or without.

**Comment by chexiongsheng** [comment]
Due to the differences in the richness of types between C# and TypeScript/JavaScript, multiple C# types may map to a single JavaScript type. A suggested solution is to encapsulate the C# API with different names to avoid overloading. For example, instead of having Foo(float), Foo(int), and Foo(short), you can encapsulate them as Foo_float(float), Foo_Int(int), and Foo_short(short).

**Comment by raad-genies** [comment]
@chexiongsheng Is it possible to actually use branded types in TS defined globally like 

```
 declare type float = number & { [floatBrand]: void };
 declare type int = number & { [intBrand]: void };
```

And include the alias info in the js call info? This would allow us to at least do type checking without having to resort to weird overload naming specially if you have some APIs that are engine specific.

**Comment by chexiongsheng** [comment]
What are the declarations of floatBrand and intBrand?

**Comment by raad-genies** [comment]
Unique symbol

**Comment by chexiongsheng** [comment]
How to call these apis? below code compile fail: error TS2345: Argument of type 'number' is not assignable to parameter of type 'int'.
  Type 'number' is not assignable to type '{ [intBrand]: void; }'.

```
declare type float = number & { [floatBrand]: void };
declare type int = number & { [intBrand]: void };

function foo(a:int) {

}
foo(100);
```

**Comment by chexiongsheng** [comment]
Not all overloads require renaming; only a small number of APIs that distinguish between overload types that happen to map to a specific TypeScript type need to be renamed.

**Comment by raad-genies** [comment]
Yeah no worries. I followed a different route to fix this since we have our own compiler, the approach we're using allows us to replace overload calls by inferring the type based on literal or if the type matches this way it's seamless to devs. The other approach I was thinking of is just being able to pass aliases in the js call info. 

Appreciate the help!
